### DISCUSSION TITLE
OnMarketData  execution throttle?

### DISCUSSION THREAD

**Discussion by lavalampmj** [MicroTrendsLtd/NinjaTrader8#76] [Q&A]
This code throttles the execution of OnMarketData to once per second or more in real-time , it will not return in Playback or Historical.

 ` if (DateTime.Now < onMarketDataTimeNextAllowed) return;
  onMarketDataTimeNextAllowed = DateTime.Now.AddSeconds(1);
`

Processing of q is well throttled with 

                if (IsTEQOnMarketData && Now >= tEQNextTimeValid)
                {
                    tEQNextTimeValid = Now.AddSeconds(TEQTimerInterval);
                    if (tracing)
                        Print("OnMarketData >> TEQ");

                    //no lock needed here on TEQ - it can be rouhgly accurate ProcessTradeEventQueue will work it out
                    if (IsTradeWorkFlowReady() && TEQ.Count > 0)
                        TriggerCustomEvent(ProcessTradeEventQueue, null);

                    this.inOnMarketData = false;
                    return;

                }

Why is it necessary, is it still desirable given the above q code?

**Comment by MicroTrendsTom** [comment]
ok i misread this a bit and had to revisit the code

Property IsRealtimeTradingUseQueue  is used the timed Q mode
This was a relic from the migrated code base - the idea was it allows a trader to cater for fast signals and to not tie up the system
it will instead execute the most recent signal last in and purge the prior for a given time frame.

TEQOnMarketDataEnable();  switches on the q pattern on - when a signal arrives

TEQTimerInterval - the timer interval can be greater than 1 up to 5

if (DateTime.Now < onMarketDataTimeNextAllowed) return; onMarketDataTimeNextAllowed = DateTime.Now.AddSeconds(1);
so the procedurally prior block of 1 second is ok above  but yes i get it... 
seems like a duplicate/redundant call but it depends on the timer interval for the q.  TEQTimerInterval  it could be 5 seconds for example.
so  a stream of fast signals cold be executed after a min of  1 to 5 seconds etc

As it is not breaking anything and probably you are the only one to see it id say its ok as it stands -

 unless you want to execute the q on 1 second or lower then a change to the flow or intervals to milliseconds is required to allow etc  
or via usage of bool comparison to flow down etc  eg. and Or with 
if (DateTime.Now < onMarketDataTimeNextAllowed) can be combined with 
 IsRealtimeTradingUseQueue   IsTEQOnMarketData  allow flow to drop down etc
 
 
This cold also be changed to allow it to drop down quicker etc for example via a property 
var interval =250
if (DateTime.Now < onMarketDataTimeNextAllowed) return; onMarketDataTimeNextAllowed = DateTime.Now.AddMilliSeconds(interval);
